### ISSUE TITLE
Implement Validation Process for Verifying Key Contract

### ISSUE THREAD

**Issue by sifnoc** [summa-dev/summa-solvency#299]
The user who wants to verify their proofs on the Summa contract, can access the verifying key contract registered within the Summa contract. However, there is currently no way to ensure that the bytes correctly represent polynomials, permutations, or fixed commitments. As a result, users have no choice but to trust that the proof generated from the publicly known circuit and its verifying keys is correct, without any means to independently verify the verifying key.

**Objective:** We need a validation process for the verifying key contract that includes commitments, key digest and others.

**Feature Description:**
This issue proposes implementing a validation process for the `vk_digest` from the registered verifying key contract in the Summa contract, allowing it to be compared with the `vk_digest` generated by the user.

- The validation process should be based on the UnivariateGrandSum circuit known from the GitHub repo.
- To initiate the circuit, the `N_BYTES` and `N_CURRENCIES` parameters are required, either by reading the verifying key contract or as provided by the prover.
- The `vk_digest` is a hash of the entire text of parameters, which is initiated with the circuit and parameters. The hash process is detailed in this repository - [halo2-solidity-verifier](https://github.com/summa-dev/halo2-solidity-verifier).
- The implementation should prioritize UX, ensuring that the comparison process is straightforward and user-friendly.

**Additional Context:**

- The verifying key contracts consist of multiple bytes in an assembly block, as shown below:
  ```solidity
      constructor() {
          assembly {
              mstore(0x0000, 0x00996ed64113d5d86f42caab0f3d25466b0856015472de4bb809f168aabc2567) // vk_digest
              mstore(0x0020, 0x0000000000000000000000000000000000000000000000000000000000000011) // k
              mstore(0x0040, 0x30643640b9f82f90e83b698e5ea6179c7c05542e859533b48b9953a2f5360801) // n_inv
              mstore(0x0060, 0x304cd1e79cfa5b0f054e981a27ed7706e7ea6b06a7f266ef8db819c179c2c3ea) // omega
  ```

- This issue arose while considering this related issue: https://github.com/zBlock-2/summa-solvency/issues/10

**Comment by protocolwhisper** [comment]
Hi👋🏼 correct me if I'm wrong the scope of this issue is to build the circuit that do the comparison?

**Comment by sifnoc** [comment]
> Hi, correct me if I'm wrong the scope of this issue is to build the circuit that do the comparison?

No, not for the circuit. However, it could be a solidity contract if it's possible to evaluate the `vk_digest` from the given parameters.

The scope could also be extended to the entire verification process. The verification function logic already exists here:
https://github.com/summa-dev/halo2-solidity-verifier/blob/d33972567f83f3218257b286b541ad97ba32928c/src/codegen.rs#L169

**Comment by a659871** [comment]
Hi, 

I am Core program student. Can U assign this issue to me?
Thanks!

**Comment by jkktom** [comment]
I'm interested in the issue. Can I contribute?

**Comment by sifnoc** [comment]
I saw your PR (https://github.com/summa-dev/summa-solvency/pull/301) @protocolwhisper.

First of all, I apologize for not being specific enough in the initial issue description. I'd like to provide more detailed notes on the verification process to clarify things.

1. The user who wants to verify the verifying key contract should generate it from the verifier parameters, as shown here:
   https://github.com/summa-dev/summa-solvency/blob/f11b1fe194cdde49d8d38030b45851c6829bb859/prover/bin/gen_verifier.rs#L38-L39
   - Assume that the SRS file is publicly available, like this: https://github.com/han0110/halo2-kzg-srs, and that the K parameter is also known.

2. Users can generate the verifying key by themselves without requiring any input data, as demonstrated here:
   https://github.com/summa-dev/summa-solvency/blob/f11b1fe194cdde49d8d38030b45851c6829bb859/prover/bin/gen_verifier.rs#L60

3. The user should be able to compare all the elements in the verifying key they generated in the previous step with the one deployed on the EVM, specifically the contract deployed with `Summa.sol`.
   - Assume that the user already knows the address of the verifier contract.

One key point: In V1 (master branch), the verification key is located inside the InclusionVerifier, as seen here:
   https://github.com/summa-dev/summa-solvency/blob/23587adcbd4f7c46371d2859a93bc8c5fa71f155/contracts/src/InclusionVerifier.sol#L216-L272
However, in the V2 (v2 branch), the inclusion verifier contract has been separated, as shown here: https://github.com/summa-dev/summa-solvency/blob/v2/contracts/src/VerifyingKey.sol.

Therefore, I recommend modifying the V1 verifier contract generator to use the `render_separately` method, as was done in V2. This would make it more convenient to access the parameter values.

**Comment by 2hyuk** [comment]
Hello! can i also contribute to this issue?

**Comment by bezze** [comment]
Hi! I've been following this issue for a while now, I've opened a PR to modify the contract templates to add a getter method por the verifying key. https://github.com/summa-dev/halo2-solidity-verifier/pull/2
